modules/messages/tests/server/message-to-influx.server.service.tests.js

```diff
@@ -0,0 +1,201 @@
+    beforeEach(function () {
+      message1to2.userFrom = user1._id;
+      message1to2.userTo = user2._id;
+      message2to1.userFrom = user2._id;
+      message2to1.userTo = user1._id;
+      shortMessage.userFrom = user1._id;
+      shortMessage.userTo = user2._id;
+      longMessage.userFrom = user1._id;
+      longMessage.userTo = user2._id;
+    });
+
+    context('The first message', testKeyAndValue([], message1to2, 'tag',
+      'position', 'first'));
+
+    context('The first reply', testKeyAndValue([message2to1], message1to2, 'tag',
+      'position', 'first_reply'));
+
+    context('The other message', testKeyAndValue([message2to1, message1to2],
+      message1to2, 'tag', 'position', 'other'));
+
+    context('The short message', testKeyAndValue([], shortMessage, 'tag',
+      'msgLengthType', 'short'));
+
+    context('The long message', testKeyAndValue([], longMessage, 'tag',
+      'msgLengthType', 'long'));
+
+    // too lazy to write test for the actual amount of reply time
+    context('The first reply', testKeyAndValue([message2to1], message1to2,
+      'field', 'replyTime'));
+
+    context('Other messages', testKeyAndValue([], message1to2,
+      'field', 'replyTime', -1));
+
+    context('Any message (correct length number)', testKeyAndValue([], longMessage, 'field',
+      'msgLength', longMessage.content.length));
+
+    context('All messages', testKeyAndValue([], message1to2,
+      'field', 'time'));
+
+    /**
+     * the preparation function for testing keys and values of the process
+     * function
+     * @param {Object[]} messagesBefore - array of object that would go to 'new
+     * Message(param) before saving to mongoDB via mongoose'
+     * @param {string} messagesBefore[].content - a message content
+     * @param {string|Object} messagesBefore[].userFrom - message sender's id
+     * (id in form of string or idObject (whatever mongoose accepts))
+     * @param {string|Object} messagesBefore[].userTo - message receiver's id
+     * (id in form of string or idObject (whatever mongoose accepts))
+     * @param {Object} messageToProcess - object that would go to new Message()
+     * @param {string} fieldOrTag - do we test field or tag? provide 'field' or
+     * 'tag'
+     * @param {string} key - field or tag key as will be saved to influxDB
+     * @param {mixed} value - field or tag value as will be saved to influxDB
+     * if not present, we test for the presence of the field only
+     * @returns {function} - the returned function is ready to be a second
+     * parameter of context();
+     */
+    function testKeyAndValue(messagesBefore, messageToProcess, fieldOrTag, key, value) {
+      // we wish to check whether the value is provided
+      var arglen = arguments.length;
+
+      return function () {
+        // create some initial message(s)
+        beforeEach(function (done) {
+          async.eachSeries(messagesBefore, saveMessage, done);
```

Comment:
In general, I would recommend defining the function before invoking it. The code here actually works because of javascript's "function hoisting", but that is not 100% reliable and I've seen cases in the wild where invocation before definition will cause an exception.